src/task/blocking.rs

```diff
@@ -45,20 +78,84 @@ lazy_static! {
         let (sender, receiver) = bounded(0);
         Pool { sender, receiver }
     };
+
+    // Pool task frequency calculation variables
+    static ref FREQ_QUEUE: Arc<Mutex<VecDeque<u64>>> = {
+        Arc::new(Mutex::new(VecDeque::with_capacity(FREQUENCY_QUEUE_SIZE + 1)))
+    };
+
+    // Pool size variable
+    static ref POOL_SIZE: Arc<Mutex<u64>> = Arc::new(Mutex::new(LOW_WATERMARK));
 }
 
-// Create up to MAX_THREADS dynamic blocking task worker threads.
-// Dynamic threads will terminate themselves if they don't
-// receive any work after between one and ten seconds.
-fn maybe_create_another_blocking_thread() {
-    // We use a `Relaxed` atomic operation because
-    // it's just a heuristic, and would not lose correctness
-    // even if it's random.
-    let workers = DYNAMIC_THREAD_COUNT.load(Ordering::Relaxed);
-    if workers >= MAX_THREADS {
-        return;
+// Gets the current pool size
+// Used for pool size boundary checking in pool manager
+fn get_current_pool_size() -> u64 {
+    let current_arc = POOL_SIZE.clone();
+    let current_pool_size = *current_arc.lock().unwrap();
+    LOW_WATERMARK.max(current_pool_size)
+}
+
+// Adaptive pool scaling function
+//
+// This allows to spawn new threads to make room for incoming task pressure.
+// Works in the background detached from the pool system and scales up the pool based
+// on the request rate.
+//
+// It uses frequency based calculation to define work. Utilizing average processing rate.
+fn scale_pool() {
```

Comment:
In another context there was some discussion around adding tracing (using log's `trace!` macro) to async-std. This functions seems like a good place for that, too.

Reply:
I just know that the idea was thrown around, at it was probably an opt-in feature, too. @stjepang or @yoshuawuyts would know more :)

Reply:
`log` now has kv support, which we're already using in other places too. `tracing` can pick these up too, so using ` log` is probably the best choice here overall (:

Reply:
I think we should get the `kv` support in `tracing-log` soon-ish? However I don't believe that `log` supports the `span`-like macros, so async-std might need to have its own, internal `span`-like macro and dispatch conditionally to async-log or tracing.

 I haven't considered how the differences between [async-log's `span`](https://docs.rs/async-log/2.0.0/async_log/macro.span.html) and [`tracing`'s span](https://docs.rs/tracing/0.1.7/tracing/macro.span.html) can be bridged, but I think it's feasible if `tracing'`s `span!().enter()` is used to mimic async-log's nested `span` block.

Reply:
@davidbarsky we could use [`async-log::span`](https://docs.rs/async-log/2.0.0/async_log/macro.span.html) for this, probably. Though there's still a few open issues on the repo, the interface wouldn't change.

Reply:
@yoshuawuyts Yep! I think we're on the same page about wanting to bridge the two libraries. 

I'm spitballing _one_ mechanism on bridging that gap, which consists of using mutually exclusive feature flags in a `build.rs` to dispatch to either `async-log` or `tracing`. This would entail using an `async-std`-specific `span` macro until the gaps—if any exist—between `async-std`'s `span` and `tracing`'s `span` are closed.

I'm sorry if what I communicated wasn't clear!